sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusProcessor.cs

```diff
@@ -835,5 +862,103 @@ public async ValueTask DisposeAsync()
+        /// </summary>
+        /// <param name="maxConcurrentCalls">The new max concurrent calls value. This will be reflected in the <see cref="ServiceBusProcessor.MaxConcurrentCalls"/>
+        /// property.</param>
+        public void UpdateConcurrency(int maxConcurrentCalls)
+        {
+            Argument.AssertAtLeast(maxConcurrentCalls, 1, nameof(maxConcurrentCalls));
+            lock (_syncLock)
+            {
+                ReconcileConcurrency(maxConcurrentCalls);
+
+                _maxConcurrentCalls = maxConcurrentCalls;
+            }
+        }
+
+        internal void UpdateConcurrency(int maxConcurrentSessions, int maxConcurrentCallsPerSession)
+        {
+            Argument.AssertAtLeast(maxConcurrentSessions, 1, nameof(maxConcurrentSessions));
+            Argument.AssertAtLeast(maxConcurrentCallsPerSession, 1, nameof(maxConcurrentCallsPerSession));
+
+            lock (_syncLock)
+            {
+                var newConcurrency = _sessionIds.Length > 0
+                    ? Math.Min(_sessionIds.Length, maxConcurrentSessions)
+                    : maxConcurrentSessions * maxConcurrentCallsPerSession;
+
+                ReconcileConcurrency(newConcurrency, maxConcurrentSessions);
+
+                _maxConcurrentSessions = maxConcurrentSessions;
+                _maxConcurrentCallsPerSession = maxConcurrentCallsPerSession;
+                _maxConcurrentCalls = newConcurrency;
+            }
+        }
+
+        private void ReconcileConcurrency(int newConcurrency, int newMaxSessions = default)
+        {
+            int diff = newConcurrency - _maxConcurrentCalls;
+            // increasing concurrency
+            if (diff > 0)
+            {
+                if (IsSessionProcessor)
+                {
+                    var diffSessions = newMaxSessions - _maxConcurrentSessions;
+                    if (diffSessions > 0 && _sessionIds.Length == 0)
+                    {
+                        for (int i = 0; i < diffSessions; i++)
+                        {
+                            _receiverManagers.Add(
+                                new SessionReceiverManager(
+                                    _sessionProcessor,
+                                    null,
+                                    MaxConcurrentAcceptSessionsSemaphore,
+                                    _scopeFactory,
+                                    _maxConcurrentCallsPerSession,
+                                    KeepOpenOnReceiveTimeout));
+                        }
+                    }
+                }
+                _messageHandlerSemaphore.Release(diff);
+            }
+            // decreasing concurrency
+            else if (diff < 0)
+            {
+                // limit the number of new tasks that can spawn to newly specified concurrency
+                for (int i = 0; i < Math.Abs(diff); i++)
```

Comment:
```suggestion
                var diffLimit = Math.Abs(diff);
                for (int i = 0; i < diffLimit; i++)
```

Unsure if this would be optimized away with the current compiler generation, but it may be useful to hoist to avoid the `Abs` call on each tick, just in case...